chain/lf3/ec.go

```diff
@@ -219,3 +203,36 @@ func (ec *ecWrapper) getPowerTableLotusTSK(ctx context.Context, tsk types.TipSet
 	sort.Sort(powerEntries)
 	return powerEntries, nil
 }
+
+func (ec *ecWrapper) Finalize(ctx context.Context, key gpbft.TipSetKey) error {
+	ts, err := ec.getTipSetFromF3TSK(ctx, key)
+	if err != nil {
+		return err
+	}
+	if err = ec.ChainStore.SetCheckpoint(ctx, ts); err != nil {
```

Comment:
Looking at the current `SetCheckpoint` code, we have a few issues:

1. `SetCheckpoint` won't sync. I think we need to check if we have the tipset and, if not, call `Sync` (via the syncer) first. I.e., if `getTipSEtFromF3TSK` fails, we need to sync.
2. `SetCheckpoint` will (currently) refuse to work if it's older than finality. It claims that doing so would be "too slow but... tbh, I think we should just re-write that code to call `GetTipsetByHeight`. Then it'll get to use the tipset cache/index and should be pretty fast. Otherwise, we'll have problems when bootstrapping F3: if bootstrap fails for a few rounds then tries to agree on base, it'll fail to finalize base because base will be more than finality epochs in the past.

Reply:
Further to 1:1 discussion, used `Syncer.SyncCheckpoint` to work around the potential need to sync missing tipsets and captured a separate issue to improve checkpinting performance by using `GetTipsetByHeight` [here](https://github.com/filecoin-project/lotus/issues/12475). 🙏